### DISCUSSION TITLE
Dead simple UI layout "Rectcut"

### DISCUSSION THREAD

**Discussion by mambastudio** [palexdev/MaterialFX#146] [Archived]
Great work on MaterialFX. Love the detail work in it.

This is just a general discussion, but wanted to point out, for your interest, there is a cool concept here on designing simple UI layout framework called [Rectcut](https://halt.software/dead-simple-layouts/)! The code seems quite simple, but the endless possibilities of designing UI layouts is quite astonishing. Apparently, it seems it's possible to design even a flexbox layout using it. Probably an idea to increase your arsenal in MaterialFX if it's of interests.

**Comment by palexdev** [comment]
Thanks!

Interesting project but I don't think it's relevant to JavaFX as I think layouts are quite powerful, on the page I see nothing you can't do with JavaFX as of now
Maybe just the flex box, but there should be a library for JavaFX, I think I saw it on GitHub some time ago

**Comment by palexdev** [comment]
@mambastudio Well, I guess it's never too late to change your mind. Two years later I ported the Rectcut layout concept to Java, [here](https://github.com/palexdev/rectcut-java)

**Reply by mambastudio** [reply]
Great!!! Hahaha. Yeah. Looks great. Have you tried using records because of the immutability nature of the Rect? Could be a very wonderful candidate.

**Reply by palexdev** [reply]
> Great!!! Hahaha. Yeah. Looks great. Have you tried using records because of the immutability nature of the Rect? Could be a very wonderful candidate.

From my understanding, the Rect is not immutable. For example, let's say you have an area of 40x40, and you cut 10px at the top. You end up having two areas:
- The starting area but the minY is now 10, not 0
- The cut area, which is: (0, 0, 40, 10) -> (minX, minY, maxX, maxY)

**Reply by mambastudio** [reply]
Oh yeah. You're absolutely right. I haven't looked at it for a long time. I don't know why in my mind I've always thought it was immutable, yet mutability is the core principle of its design, specifically modifying and updating the parent layout.
